### PR TITLE
[FIX] web_editor: fix colorpicker of the transparent header

### PR DESCRIPTION
**Pull request by bvr-odoo**
Since this commit [1], 3 bugs have appeared with the colorpicker
of the transparent header and 1 bug with the colorpicker of color
combinations presets.

1rst bug fix was moved to https://github.com/odoo/odoo/pull/134202

Steps to reproduce the second bug:
- Go to Website edit mode.
- Click on the header.
- Choose "Over The content" for the "Header Position" option of the
header.
- Open the colorpicker (below the "Header Position" selector).
- Bug: we shouldn't be able to choose anything other than a gray.

This bug occurs because, before the conversion to owl, the colorpicker
widget was added conditionally in JavaScript (so the widget was not
included if the option "custom" color was excluded). However, in the
current version, the color picker is inside the
'web_editor.ColorPalette' XML without any condition to display it or
not.

Steps to reproduce the 3rd bug:
- Go to Website edit mode.
- Click on the header.
- Choose "Over The content" for the "Header Position" option of the
header.
- Change the color of the header by picking a color in the "Background"
sub option (below the "Header Position" selector).
- Save and Edit.
- Go with your mouse over the different colors of the colorpicker (at
least 2).
- Bug: an infinite loader appears on the iframe and prevents from
editing the background color.

Steps to reproduce the color combinations bug:
- Go to Website edit mode.
- Drop a Call to action snippet.
- Change the background color and try to select another preset.
- Bug: the color of the text and buttons is updated, but not the
background color.

Both of these 2 bugs happen after commit [1] because the colorpicker
widget is started before the ColorPalette component is attached to the
DOM (which only happens once the user opens the picker). However, the
colorNames are only set after the setup() of the ColorPalette.
Initializing the colorNames with possible default values is thus
necessary to avoid bugs on `_computeWidgetState()`.

[1]: https://github.com/odoo/odoo/commit/d7245d2abf528d093226c80e40975e63d61e8997
[documentation]: https://github.com/odoo/owl/blob/master/doc/reference/props.md#good-practices

task-3440939

### CODE REVIEW

**Comment by qsm-odoo** [comment]
@robodoo delegate=loco-odoo 
@loco-odoo

**Comment by loco-odoo** [comment]
@bvr-odoo, I think I found a bug concerning the colorpicker of the transparent header:
-Go to Website in edit mode.
-Click on the header.
-Choose "Over The content" for the "Header Position" option of the header.
-Change the color of the header by picking a color in the "Background" sub option (below the "Header Position" selector).
- Save and Edit.
- Go with your mouse over the different colors of the colorpicker.

-> Infinite loading.

File: addons/web_editor/static/src/js/editor/snippets.options.js
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:1422]: This condition was wrongly changed by NBY in the problematic commit... why can't it be just restored to first render the colorpalette in a fragment if needed? I'd like to avoid the use of `getParent()` and having to implement this getColorNames function

File: addons/web_editor/static/src/js/wysiwyg/widgets/color_palette.js
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/wysiwyg/widgets/color_palette.js:330]: Commit message says:

> the color picker is inside the 'web_editor.ColorPalette' XML without any condition to display it

Why can't a condition be added on the XML then? Won't it be prettier?

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
I split the bug fix in 3 commits, it's better. I extracted the first one to https://github.com/odoo/odoo/pull/134202 to be merged.

I am adding comments for the remaining two here :+1:

File: addons/web_editor/static/src/js/wysiwyg/widgets/color_palette.js
Comment (bvr-odoo) [reply addons/web_editor/static/src/js/wysiwyg/widgets/color_palette.js]: > Why can't a condition be added on the XML then? Won't it be prettier?

Indeed, thanks. I made the change

File: addons/web_editor/static/src/js/editor/snippets.options.js
Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:1422]: @bvr-odoo @robinlej I forgot about this. I still cannot merge this PR like that: if we don't have a solution, TODO comments should explain what's going on so we can investigate later. I see there is still this code right now:
```javascript
onSetColorNames: (colorNames) => {
    this.colorPaletteColorNames = colorNames
},
```
So this seems like a duplicated code to retrieves the color names, right? :thinking: (is this even needed since your hack with `getParent` and the new function? (if your function is supposed to be the new way to retrieve colors, do we still need that old one? (which is also strange and ugly somehow ^^)) :thinking: 

Side comment: the author of both remaining commits is BVR, should you be a co-author @robinlej ? Was not sure who to ping on this :)

Comment (robinlej) [reply addons/web_editor/static/src/js/editor/snippets.options.js:1422]: @qsm-odoo Not sure what type of comments you're expecting, I thought the one I wrote last time was enough. ^^' I've reworded it a bit and added a "TODO" to let it stand out, let me know if you think there should still be more to it / other comments.

I removed the `onSetColorNames` prop inside the snippets.options (it should work indeed, and as far as I can tell after testing, it does), but left it in general for the ColorPalette component (it's used in the web_editor code for more specific uses, and I guess this should anyway be the way to go once the options are truly converted to OWL).

This all being said, there's one test failing since your rebase 3 weeks ago in mass_mailing... It does work on local, so I'm investigating.

(I'm indeed author of this commit. I've added the line.)

Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:1422]: > Not sure what type of comments you're expecting

Basically just adding "TODO" in the comment and not having multiple ways the colors are fetched is what I expected, so what you did :wink: 

> (I'm indeed author of this commit. I've added the line.)

If you are the only one, you can also change the author entirely (`--author=`)

> I'm investigating.

Ping me once ok, thanks :+1:

**Comment by robinlej** [comment]
@qsm-odoo I finally found a way to have the test pass. :) It should be all good now.

Quick explanation: the test (mass_mailing_code_view_tour) doesn't actually work (I've forwarded the info to the editor team which made the test).
The consequence of it is that it opens and closes the editor so quickly the wysiwyg is started > mounted > destroyed before it's had the time to fetch the colorpickerTemplate. So when it tries to call the getColorPickerTemplateService, it fails. Solution: override `getColorPickerTemplateService()` at the willDestroy lifecycle step.

**Review by qsm-odoo — APPROVED** [review]
Let's go with that, we'll find better in master later. Also, for the next bugs, let's try to add some tests with the fixes :pray: Those are important flows (at least the one with being able to switch color combinations), they should not break silently. But let's focus on the freeze for now.

@robodoo rebase-ff r+

**Pull request closed without merging**
